File: sample-chat-obj-c/sample-chat/Components/ChatScreen/Views/AttachmentBar/AttachmentBar.m
//
//  AttachmentBar.m
//  samplechat
//
//  Created by Injoit on 2/25/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

#import "AttachmentBar.h"
#import "UIImage+fixOrientation.h"

@interface AttachmentBar()
@property (strong, nonatomic) UIImageView *imageView;
@property (strong, nonatomic) UILabel *progressLabel;
@property (strong, nonatomic) UIButton *cancelButton;
@end

@implementation AttachmentBar
    
//MARK: - Life Cycle
- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self setupView];
    }
    return self;
}
    
//MARK - Setup
- (void)setupView {
    self.imageView = [[UIImageView alloc] init];
    [self addSubview:self.imageView];
    
    self.progressLabel = [[UILabel alloc] init];
    [self.imageView addSubview:self.progressLabel];
    
    self.cancelButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [self addSubview:self.cancelButton];
    
    self.imageView.translatesAutoresizingMaskIntoConstraints = NO;
    self.progressLabel.translatesAutoresizingMaskIntoConstraints = NO;
    self.cancelButton.translatesAutoresizingMaskIntoConstraints = NO;
    
    [self.imageView.leftAnchor constraintEqualToAnchor:self.leftAnchor constant:20.0f].active = YES;
    [self.imageView.topAnchor constraintEqualToAnchor:self.topAnchor constant:10.0f].active = YES;
    [self.imageView.widthAnchor constraintEqualToConstant:80.0f].active = YES;
    [self.imageView.heightAnchor constraintEqualToConstant:80.0f].active = YES;
    self.imageView.contentMode = UIViewContentModeScaleAspectFill;
    self.imageView.layer.cornerRadius = 8.0f;
    self.imageView.layer.masksToBounds = YES;
    
    [self.progressLabel.leftAnchor constraintEqualToAnchor:self.imageView.leftAnchor].active = YES;
    [self.progressLabel.rightAnchor constraintEqualToAnchor:self.imageView.rightAnchor].active = YES;
    [self.progressLabel.centerYAnchor constraintEqualToAnchor:self.imageView.centerYAnchor].active = YES;
    [self.progressLabel.heightAnchor constraintEqualToConstant:40.0f].active = YES;
    self.progressLabel.textAlignment = NSTextAlignmentCenter;
    self.progressLabel.font = [UIFont systemFontOfSize:16.0f];
    self.progressLabel.textColor = [UIColor whiteColor];
    
    [self.cancelButton.rightAnchor constraintEqualToAnchor:self.imageView.rightAnchor constant:-4.0f].active = YES;
    [self.cancelButton.topAnchor constraintEqualToAnchor:self.imageView.topAnchor constant:2.0f].active = YES;
    [self.cancelButton.heightAnchor constraintEqualToConstant:24.0f].active = YES;
    [self.cancelButton.heightAnchor constraintEqualToConstant:24.0f].active = YES;
    [self.cancelButton setImage:[UIImage imageNamed:@"ic_cancel"] forState:UIControlStateNormal];
    [self.cancelButton setTintColor:[UIColor whiteColor]];
    [self.cancelButton setEnabled:YES];
    [self.cancelButton setHidden:YES];
    [self.cancelButton addTarget:self action:@selector(cancelButtonPressed:) forControlEvents:UIControlEventTouchUpInside];
    
}

- (void)cancelButtonPressed:(UIButton *)sender {
    self.cancelButton.hidden = YES;
    if ([self.delegate respondsToSelector:@selector(attachmentBar:didTapCancelButton:)]) {
        [self.delegate attachmentBar:self didTapCancelButton:sender];
    }
}

- (void)updateLoadingProgress:(CGFloat)progress {
    if (progress > 0.0) {
        self.progressLabel.hidden = NO;
        self.progressLabel.text = [NSString stringWithFormat:@"%2.0f %%", progress * 100.0];
    }
        if (progress > 0.99) {
            self.progressLabel.hidden = YES;
            self.cancelButton.hidden = NO;
        }
}

/**
 *  This method is called when the user finishes picking attachment image.
 *
 *  @param image    image that was picked by user
 */
- (void)uploadAttachmentImage:(UIImage *)image pickerControllerSourceType:(UIImagePickerControllerSourceType)sourceType {
    
    self.imageView.image = image;
    self.cancelButton.hidden = YES;
    
    __weak typeof(self)weakSelf = self;
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        __typeof(weakSelf)strongSelf = weakSelf;
        
        UIImage *newImage = image;
        if (sourceType == UIImagePickerControllerSourceTypeCamera) {
            newImage = [newImage fixOrientation];
        }
        
        CGFloat largestSide = image.size.width > image.size.height ? image.size.width : image.size.height;
        CGFloat scaleCoefficient = largestSide / 560.0f;
        CGSize newSize = CGSizeMake(image.size.width / scaleCoefficient, image.size.height / scaleCoefficient);
        
        UIGraphicsBeginImageContext(newSize);
        
        [image drawInRect:(CGRect){0, 0, newSize.width, newSize.height}];
        UIImage *resizedImage = UIGraphicsGetImageFromCurrentImageContext();
        UIGraphicsEndImageContext();
        if (resizedImage == nil) {
            return;
        }
        NSData *imageData = UIImagePNGRepresentation(resizedImage);
        // Sending attachment to the dialog.
        dispatch_async(dispatch_get_main_queue(), ^{
            
            [QBRequest TUploadFile:imageData
                          fileName:@"test.png"
                       contentType:@"image/png"
                          isPublic:NO
                      successBlock:^(QBResponse * _Nonnull response, QBCBlob * _Nonnull uploadedBlob) {
                          
                          QBChatAttachment *attachment = [[QBChatAttachment alloc] init];
                          
                          attachment.url = uploadedBlob.privateUrl ? uploadedBlob.privateUrl : uploadedBlob.publicUrl;
                          
                          attachment.ID = uploadedBlob.UID;
                          attachment.name = uploadedBlob.name;
                          attachment.type = @"image";
                          
                          if ([strongSelf.delegate respondsToSelector:@selector(attachmentBar:didUpLoadAttachment:)]) {
                              [strongSelf.delegate attachmentBar:strongSelf didUpLoadAttachment:attachment];
                          }
                          
                      } statusBlock:^(QBRequest * _Nonnull request, QBRequestStatus * _Nonnull status) {
                          
                          CGFloat progress = status.percentOfCompletion;
                          [strongSelf updateLoadingProgress:progress];
                          
                      } errorBlock:^(QBResponse * _Nonnull response) {
                          if ([strongSelf.delegate respondsToSelector:@selector(attachmentBarFailedUpLoadImage:)]) {
                              [strongSelf.delegate attachmentBarFailedUpLoadImage:strongSelf];
                          }
                          
                      }];
        });
    });
}

@end

File: sample-chat-swift/sample-chat-swift/Classes/Chat/CustomUI/AttachmentBar/AttachmentBar.swift
//
//  AttachmentBar.swift
//  sample-chat-swift
//
//  Created by Injoit on 1/28/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

import UIKit

protocol AttachmentBarDelegate: class {
    func attachmentBarFailedUpLoadImage(_ attachmentBar: AttachmentBar);
    func attachmentBar(_ attachmentBar: AttachmentBar, didUpLoadAttachment  attachment: QBChatAttachment)
    func attachmentBar(_ attachmentBar: AttachmentBar, didTapCancelButton: UIButton)
}

class AttachmentBar: UIView {
    
    //MARK: - Properties
    weak var delegate: AttachmentBarDelegate?
    
    lazy var imageView: UIImageView = {
        let imageView = UIImageView()
        imageView.setRoundedView(cornerRadius: 8.0)
        addSubview(imageView)
        imageView.translatesAutoresizingMaskIntoConstraints = false
        imageView.leftAnchor.constraint(equalTo: leftAnchor, constant: 20.0).isActive = true
        imageView.topAnchor.constraint(equalTo: topAnchor, constant: 10.0).isActive = true
        imageView.widthAnchor.constraint(equalToConstant: 80.0).isActive = true
        imageView.heightAnchor.constraint(equalToConstant: 80.0).isActive = true
        imageView.contentMode = .scaleAspectFill
        return imageView
    }()
    
    lazy var progressLabel: UILabel = {
        let progressLabel = UILabel()
        progressLabel.textAlignment = .center
        progressLabel.font = .systemFont(ofSize: 16.0)
        progressLabel.textColor = .white
        imageView.addSubview(progressLabel)
        progressLabel.translatesAutoresizingMaskIntoConstraints = false
        progressLabel.leftAnchor.constraint(equalTo: imageView.leftAnchor).isActive = true
        progressLabel.rightAnchor.constraint(equalTo: imageView.rightAnchor).isActive = true
        progressLabel.centerYAnchor.constraint(equalTo: imageView.centerYAnchor).isActive = true
        progressLabel.heightAnchor.constraint(equalToConstant: 40.0).isActive = true
        return progressLabel
    }()
    
    lazy var cancelButton: UIButton = {
        let cancelButton = UIButton(type: .system)
        cancelButton.setImage(UIImage(named: "ic_cancel"), for: .normal)
        cancelButton.addTarget(self, action: #selector(cancelButtonPressed(_:)), for: .touchUpInside)
        cancelButton.tintColor = .white
        cancelButton.isEnabled = true
        addSubview(cancelButton)
        cancelButton.translatesAutoresizingMaskIntoConstraints = false
        cancelButton.rightAnchor.constraint(equalTo: imageView.rightAnchor, constant: -2.0).isActive = true
        cancelButton.topAnchor.constraint(equalTo: imageView.topAnchor, constant: 2.0).isActive = true
        cancelButton.widthAnchor.constraint(equalToConstant: 24.0).isActive = true
        cancelButton.heightAnchor.constraint(equalToConstant: 24.0).isActive = true
        cancelButton.addTarget(self, action: #selector(cancelButtonPressed(_:)), for: .touchUpInside)
        cancelButton.isHidden = true
        return cancelButton
    }()
    
    //MARK: - Actions
    /**
     *  This method is called when the user finishes picking attachment image.
     *
     *  @param image    image that was picked by user
     */
    func uploadAttachmentImage(_ image: UIImage, sourceType: UIImagePickerController.SourceType) {
        
        imageView.image = image;
        cancelButton.isHidden = true;
        
        DispatchQueue.global().async { [weak self] () -> Void in
            
            guard let self = self else { return }
            
            var newImage = image
            if sourceType == .camera {
                newImage = newImage.fixOrientation()
            }
            
            let largestSide = newImage.size.width > newImage.size.height ? newImage.size.width : newImage.size.height
            let scaleCoeficient = largestSide/560.0
            let newSize = CGSize(width: newImage.size.width/scaleCoeficient, height: newImage.size.height/scaleCoeficient)
            
            // create smaller image
            UIGraphicsBeginImageContext(newSize)
            
            newImage.draw(in: CGRect(x: 0, y: 0, width: newSize.width, height: newSize.height))
            let resizedImage = UIGraphicsGetImageFromCurrentImageContext()
            
            UIGraphicsEndImageContext()
            
            // Sending attachment.
            DispatchQueue.main.async(execute: {
                // sendAttachmentMessage method always firstly adds message to memory storage
                guard let imageData = resizedImage?.pngData() else {
                    return
                }
                QBRequest.tUploadFile(imageData, fileName: "test.png",
                                      contentType: "image/png", isPublic: false,
                                      successBlock: { [weak self] (response: QBResponse, uploadedBlob: QBCBlob) -> Void in
                                        guard let self = self else {
                                                return
                                        }
                                        let attachment = QBChatAttachment()
                                        if let privateUrl = uploadedBlob.privateUrl() {
                                            attachment.url = privateUrl
                                        } else if let publicUrl = uploadedBlob.publicUrl() {
                                            attachment.url = publicUrl
                                        } else {
                                            return
                                        }
                                        attachment.id = uploadedBlob.uid
                                        attachment.name = uploadedBlob.name
                                        attachment.type = "image"
                                        
                                        self.delegate?.attachmentBar(self, didUpLoadAttachment: attachment)
                                        
                    }, statusBlock: { [weak self] (request : QBRequest?, status : QBRequestStatus?) -> Void in
                        if let status = status {
                            DispatchQueue.main.async {
                                let progress = CGFloat(Float(status.percentOfCompletion))
                                self?.updateLoadingProgress(progress)
                            }
                        }
                }) { [weak self] (response : QBResponse) -> Void in
                    guard let self = self else {
                        return
                    }
                    self.delegate?.attachmentBarFailedUpLoadImage(self)

                }
            })
        }
    }
    
    
    @objc func cancelButtonPressed(_ sender: UIButton) {
        cancelButton.isHidden = true
        delegate?.attachmentBar(self, didTapCancelButton: sender)
    }
    
    func updateLoadingProgress(_ progress: CGFloat) {
        if progress > 0.0 {
            progressLabel.isHidden = false
            progressLabel.text = String(format: "%2.0f %%", progress * 100.0)
        }
        if progress > 0.99 {
            progressLabel.isHidden = true
            cancelButton.isHidden = false
        }
    }
}
